### PR TITLE
Improvements and changes to progress tracking dataclasses

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

- Move the batch attributes inside the epoch progress. This follows the structure in the loops where the batch loop is an attribute in the epoch loop. This makes it easier for users to compose custom progress dataclasses and connect them across loops.
- Implement `{,from_}state_dict` for all dataclasses
- Follows the structure in https://github.com/PyTorchLightning/pytorch-lightning/pull/8120

Necessary to unblock https://github.com/PyTorchLightning/pytorch-lightning/pull/7976

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -85,12 +84,14 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Fault-tolerant training
     * Add `{,load_}state_dict` to `ResultCollection` ([#7948](https://github.com/PyTorchLightning/pytorch-lightning/pull/7948))
-    * Checkpoint the loop results ([#7966](https://github.com/PyTorchLightning/pytorch-lightning/pull/7966))
 
 
 - Add `rank_zero_only` to `LightningModule.log` function ([#7966](https://github.com/PyTorchLightning/pytorch-lightning/pull/7966))
 
 
+- Add `metric_attribute` to `LightningModule.log` function ([#7966](https://github.com/PyTorchLightning/pytorch-lightning/pull/7966))
```
Comment (carmocca) [inline CHANGELOG.md:92]: Adding this here because forgot to do it in #7966

File: CHANGELOG.md [hunk 2]
```diff
@@ -85,12 +84,14 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Fault-tolerant training
     * Add `{,load_}state_dict` to `ResultCollection` ([#7948](https://github.com/PyTorchLightning/pytorch-lightning/pull/7948))
-    * Checkpoint the loop results ([#7966](https://github.com/PyTorchLightning/pytorch-lightning/pull/7966))
```
Comment (carmocca) [inline CHANGELOG.md:88]: This wasn't added in #7966

File: pytorch_lightning/trainer/progress.py [hunk 3]
```diff
@@ -11,12 +11,31 @@
...
+        return asdict(self)
+
+    def __setstate__(self, state: dict) -> None:
+        self.__dict__.update(state)
+
+    def state_dict(self) -> dict:
+        return self.__getstate__()
+
+    @classmethod
+    def load_state_dict(cls, state_dict: dict) -> "_DataclassStateDictMixin":
```
Comment (awaelchli) [inline pytorch_lightning/trainer/progress.py:31]: I think this is not consistent with the rest of the load_state_dict semantics where we operate on an instance.
I would like to suggest simply renaming this to `from_state_dict`.
Optionally a separate load_state_dict method on an instance could also be useful.

**Review by awaelchli — APPROVED** [review]

File: CHANGELOG.md [hunk 4]
```diff
@@ -30,9 +30,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added support for checkpointing based on a provided time interval during training ([#7515](https://github.com/PyTorchLightning/pytorch-lightning/pull/7515))
 
 
-- Added dataclasses for progress tracking (
-    [#6603](https://github.com/PyTorchLightning/pytorch-lightning/pull/6603),
-    [#7574](https://github.com/PyTorchLightning/pytorch-lightning/pull/7574))
+- Progress tracking
+  * Added dataclasses for progress tracking ([#6603](https://github.com/PyTorchLightning/pytorch-lightning/pull/6603), [#7574](https://github.com/PyTorchLightning/pytorch-lightning/pull/7574), [#8140](https://github.com/PyTorchLightning/pytorch-lightning/pull/8140))
+  * Add `{,load_}state_dict` to the progress tracking dataclasses ([#8140](https://github.com/PyTorchLightning/pytorch-lightning/pull/8140))
```
[the remark this answers is not in the document]
Comment (carmocca) [reply CHANGELOG.md:35]: Bash terminal notation

```console
echo {,load_}state_dict
state_dict load_state_dict
```

File: pytorch_lightning/trainer/progress.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/trainer/progress.py:31]: Good point. Will change it

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by justusschock — APPROVED** [review]
lgtm
